### PR TITLE
[SPARK-33482][SQL] Fix FileScan canonicalization

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?

This PR adds canonicalization to `FileScan.partitionFilters` and `FileScan.dataFilters` in `BatchScanExec` nodes.

### Why are the changes needed?

Partition filters and data filters added to `FileScan` (in https://github.com/apache/spark/pull/27112 and https://github.com/apache/spark/pull/27157) caused that canonicalized form of some `BatchScanExec` nodes don't match and this prevents some reuse possibilities.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 1]
```diff
@@ -86,7 +86,7 @@ trait FileScan extends Scan
 
   override def equals(obj: Any): Boolean = obj match {
     case f: FileScan =>
-      fileIndex == f.fileIndex && readSchema == f.readSchema
+      fileIndex == f.fileIndex && readSchema == f.readSchema &&
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:89]: This change is not related to this PR, but it looks like a `&&` is missing from here.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:89]: Nice catch. It seems that https://github.com/apache/spark/pull/27112 introduced this. 
 cc @dongjoon-hyun if we want to backport this to `3.1` and `3.0`.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -4065,6 +4066,33 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+            spark.read.parquet(path.toString).createOrReplaceTempView("t")
+            val df = sql(
+              """
+                |SELECT *
+                |FROM t AS t1
+                |JOIN t AS t2 ON t2.id = t1.id
+                |JOIN t AS t3 ON t3.id = t2.id
+                |""".stripMargin)
+            df.collect()
+            df.explain()
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4086]: do we really need these two statements?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:89]: Oh... nice catch. cc: @gengliangwang , too.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -4065,6 +4066,33 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       }
     }
   }
+
+  test("SPARK-33482: Fix FileScan canonicalization") {
+    Seq(true, false).foreach { aqe =>
+      withSQLConf(SQLConf.USE_V1_SOURCE_LIST.key -> "",
+        SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> aqe.toString) {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4073]: We need to set this config for this test purpose?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 1 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:89]: Credit goes to @bersprockets for catching this in SPARK-33482.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 2 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: No. I left `explain()` there accidentally, removed in: https://github.com/apache/spark/pull/31820/commits/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Well, reuse exchange code is very different in AQE/non-AQE paths, but I think you are right as we just need to test the canonicalization fix so I've removed this in https://github.com/apache/spark/pull/31820/commits/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1

**Review by maropu — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:89]: wow ..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 4]
```diff
@@ -49,6 +49,13 @@ case class BatchScanExec(
   }
 
   override def doCanonicalize(): BatchScanExec = {
-    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)))
+    val canonicalizedScan = scan match {
+      case s: FileScan =>
+        s.withFilters(QueryPlan.normalizePredicates(s.partitionFilters, output),
+          QueryPlan.normalizePredicates(s.dataFilters, output))
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:55]:
```suggestion
        s.withFilters(
          QueryPlan.normalizePredicates(s.partitionFilters, output),
          QueryPlan.normalizePredicates(s.dataFilters, output))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 5]
```diff
@@ -49,6 +49,13 @@ case class BatchScanExec(
...
+      case _ => scan
+    }
+    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)),
+      scan = canonicalizedScan)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:59]:
```suggestion
    this.copy(
      output = output.map(QueryPlan.normalizeExpressions(_, output)),
      scan = canonicalizedScan)
```

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -4065,6 +4066,29 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+        withTempView("t") {
+          spark.read.parquet(path.toString).createOrReplaceTempView("t")
+          val df = sql(
+            """
+              |SELECT *
+              |FROM t AS t1
+              |JOIN t AS t2 ON t2.id = t1.id
+              |JOIN t AS t3 ON t3.id = t2.id
+              |""".stripMargin)
+          df.collect()
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4083]: do we really need `df.collect()` here? shouldn't `AdaptiveSparkPlanHelper.collect()` below take care of going through query plan properly with AQE being enabled?

**Review by c21 — APPROVED** [review]
LGTM with minor test comment.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 6 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4083]: Yes, we do need to run the query first and then check the plan as this is an AQE compatible query where `ReusedExchangeExec` nodes are inserted during execution.

**Review by dongjoon-hyun — COMMENTED** [review]
This is a nice patch, @peter-toth . For a better traceability, please proceed `FileScan` change in a separate PR. It looks worth to have a new JIRA because it looks like a correctness issue. And, if possible, with a separate test case focus on `FileScan.equals`.

**Comment by dongjoon-hyun** [comment]
Also, cc @cloud-fan .

**Comment by peter-toth** [comment]
> This is a nice patch, @peter-toth . For a better traceability, please proceed `FileScan` change in a separate PR. It looks worth to have a new JIRA because it looks like a correctness issue. And, if possible, with a separate test case focus on `FileScan.equals`.

All right, reverted in https://github.com/apache/spark/pull/31820/commits/483686d98e887c278974c0ef89bf32cbb22b580d and extracted to https://github.com/apache/spark/pull/31848

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 7]
```diff
@@ -49,6 +49,15 @@ case class BatchScanExec(
   }
 
   override def doCanonicalize(): BatchScanExec = {
-    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)))
+    val canonicalizedScan = scan match {
+      case s: FileScan =>
+        s.withFilters(
+          QueryPlan.normalizePredicates(s.partitionFilters, output),
+          QueryPlan.normalizePredicates(s.dataFilters, output))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:56]: This works, but is a bit hacky as it doesn't apply to all the `Scan` implementations.

I think we should add doc in the `Scan` interface to explain how the `hashCode/equals` should be implemented.

**Comment by peter-toth** [comment]
Reviewers, I've moved the e2e test to https://github.com/apache/spark/pull/31848 and fixed the issue there by changing `FileScan.equals()`.

**Comment by maropu** [comment]
We can close this now?

**Pull request closed without merging**